include/antworld/surface.h

```diff
@@ -0,0 +1,140 @@
+// Surface
+//------------------------------------------------------------------------
+// Encapsulates a single 'surface' - geometry to be rendered in one draw call using one material
+namespace BoBRobotics
+{
+namespace AntWorld
+{
+class Surface
+{
+public:
+    Surface();
+    ~Surface();
+
+    //------------------------------------------------------------------------
+    // Public API
+    //------------------------------------------------------------------------
+    void bind(bool bindTexture = true) const;
+    void unbind(bool unbindTexture = true) const;
+    void unbindIndices() const;
+    void render(GLenum primitive = GL_TRIANGLES, GLenum indexType = GL_UNSIGNED_INT) const;
+
+    template<typename T>
+    void uploadPositions(const std::vector<T> &positions, GLint size = 3)
+    {
+        // Upload positions to buffer
+        uploadBuffer(positions, m_PositionVBO, GL_ARRAY_BUFFER, GL_STATIC_DRAW);
+
+        // Set vertex pointer and enable client state in VAO
+        glVertexPointer(size, OpenGLTypeTraits<T>::type, 0, BUFFER_OFFSET(0));
+        glEnableClientState(GL_VERTEX_ARRAY);
+
+        // Calculate number of vertices from positions
+        m_NumVertices = positions.size() / size;
+
+        // Unbind buffer
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+    }
+
+    template<typename T>
+    void uploadColours(const std::vector<T> &colours, GLint size = 3)
+    {
+        // Upload colours to buffer
+        uploadBuffer(colours, m_ColourVBO, GL_ARRAY_BUFFER, GL_STATIC_DRAW);
+
+        // Set colour pointer and enable client state in VAO
+        glColorPointer(size, OpenGLTypeTraits<T>::type, 0, BUFFER_OFFSET(0));
+        glEnableClientState(GL_COLOR_ARRAY);
+
+        // Unbind buffer
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+    }
+
+    template<typename T>
+    void uploadTexCoords(const std::vector<T> &texCoords, GLint size = 2)
+    {
+        // Upload texture coordinates to buffer
+        uploadBuffer(texCoords, m_TexCoordVBO, GL_ARRAY_BUFFER, GL_STATIC_DRAW);
+
+        // Set colour pointer and enable client state in VAO
+        glTexCoordPointer(size, OpenGLTypeTraits<T>::type, 0, BUFFER_OFFSET(0));
+        glEnableClientState(GL_TEXTURE_COORD_ARRAY);
+
+        // Unbind buffer
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+    }
+
+    template<typename T>
+    void uploadIndices(const std::vector<T> &indices)
+    {
+        // Upload indices
+        uploadBuffer(indices, m_IBO, GL_ELEMENT_ARRAY_BUFFER, GL_STATIC_DRAW);
+
+        // Cache number of indices
+        m_NumIndices = indices.size();
+
+        // **NOTE** GL_ELEMENT_ARRAY_BUFFER works subtly different from GL_ARRAY_BUFFER
+        // as it has no client state/pointer tying it to the VAO. Therefore we need to
+        // leave it bound until after we unbind the VAO - makes sense but ugly
+    }
+    
+    void setTexture(const Texture *texture){ m_Texture = texture; }
```

Comment:
Could ``texture`` be ``nullptr``? If not, maybe we should pass by reference or check with ``BOB_ASSERT``

Reply:
Yeah, that just signifies no texture

Reply:
Yeah ok. A pointer seems right then.